File: src/services/tools/web/fetchPage.ts
import { z } from 'zod';
import type { ToolDef } from '../types';
import { WebSearchService } from '../../WebSearchService';

export const fetchPageTool: ToolDef = {
  name: 'fetch_page',
  description: 'Fetch and extract the full text content of a webpage from a URL. Use when you need more detail than search snippets provide.',
  category: 'web',
  inputSchema: z.object({
    url: z.string().url().describe('The full URL to fetch (must include https://).'),
    extractAs: z.enum(['markdown', 'text']).optional().default('markdown').describe('Format to extract the page content as.'),
  }),
  execute: async ({ url, extractAs }) => {
    try {
      const result = await WebSearchService.fetchPage({ url, extractAs });
      return { content: result.content, title: result.title, url: result.url };
    } catch (error: any) {
      throw new Error(error.message || 'Could not fetch the page.');
    }
  },
};

File: agent/internal/tool/fetch_page.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func fetchPageTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "fetch_page",
			Description: "Fetch the full content of a webpage as markdown or text. Use this to read articles, documentation, or API responses.",
			Category:    "research",
			Parameters: map[string]api.ParamDef{
				"url":    {Type: "string", Description: "The URL to fetch", Required: true},
				"format": {Type: "string", Description: "Output format", Required: false, Default: "markdown", Enum: []string{"markdown", "text"}},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			url, _ := params["url"].(string)
			if url == "" {
				return nil, fmtError("url is required")
			}
			format, _ := params["format"].(string)
			if format == "" {
				format = "markdown"
			}

			return postToExpress(ctx, e.expressURL, "fetchPage", map[string]any{
				"url":       url,
				"extractAs": format,
			})
		},
	}
}
